base/src/com/google/idea/blaze/base/sync/workspace/VirtualIncludesHandler.java

```diff
@@ -188,4 +216,121 @@ public static ImmutableList<String> collectIncludeHints(
 
     return result;
   }
+
+  private static List<Path> splitExecutionPath(ExecutionRootPath executionRootPath) {
+    return Lists.newArrayList(executionRootPath.getAbsoluteOrRelativeFile().toPath().iterator());
+  }
+
+  static boolean containsVirtualInclude(ExecutionRootPath executionRootPath) {
+    return splitExecutionPath(executionRootPath).contains(VIRTUAL_INCLUDES_DIRECTORY);
+  }
+
+  /**
+   * Resolves execution root path to {@code _virtual_includes} directory to the matching workspace
+   * location
+   *
+   * @return list of resolved paths if required information is obtained from execution root path and
+   * target data or empty list if resolution has failed
+   */
+  @NotNull
+  static ImmutableList<File> resolveVirtualInclude(ExecutionRootPath executionRootPath,
+      File externalWorkspacePath,
+      WorkspacePathResolver workspacePathResolver,
+      TargetMap targetMap) {
+    TargetKey key = null;
+    try {
+      key = guessTargetKey(executionRootPath);
+    } catch (IndexOutOfBoundsException exception) {
+      // report to intellij EA
+      LOG.error(
+          "Failed to detect target from execution root path: " + executionRootPath,
+          exception);
+    }
+
+    if (key == null) {
+      return ImmutableList.of();
+    }
+
+    TargetIdeInfo info = targetMap.get(key);
+    if (info == null) {
+      return ImmutableList.of();
+    }
+
+    CIdeInfo cIdeInfo = info.getcIdeInfo();
+    if (cIdeInfo == null) {
+      return ImmutableList.of();
+    }
+
+    if (!info.getcIdeInfo().getIncludePrefix().isEmpty()) {
+      LOG.debug(
+          "_virtual_includes cannot be handled for targets with include_prefix attribute");
+      return ImmutableList.of();
+    }
+
+    String stripPrefix = info.getcIdeInfo().getStripIncludePrefix();
+    if (!stripPrefix.isEmpty()) {
+      if (stripPrefix.endsWith("/")) {
+        stripPrefix = stripPrefix.substring(0, stripPrefix.length() - 1);
+      }
+      String externalWorkspaceName = key.getLabel().externalWorkspaceName();
+      WorkspacePath stripPrefixWorkspacePath = stripPrefix.startsWith(ABSOLUTE_LABEL_PREFIX) ?
+          new WorkspacePath(stripPrefix.substring(ABSOLUTE_LABEL_PREFIX_LENGTH)) :
+          new WorkspacePath(key.getLabel().blazePackage(), stripPrefix);
+      if (key.getLabel().externalWorkspaceName() != null) {
```

Comment:
this code has a potential NPE, could we also address this issue? https://github.com/bazelbuild/intellij/pull/5969#issuecomment-1934181555